Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -164,6 +184,9 @@ class AsyncQueue {
 
   std::atomic<bool> is_operation_in_progress_;
   std::unique_ptr<Executor> executor_;
+
+  bool is_shutting_down_;
+  mutable std::mutex shut_down_mutex_;
```

Comment:
What's the reason to prefer a mutex to an atomic boolean?

Reply:
atomic boolean cannot make sure scheduling is synchronized. It is possible in a `Enqueue` method, right after the check for `is_shutting_down` and before enqueue, another thread shuts down the queue. Then we will have some operations scheduled after the queue is shutdown, which might leads to strange things to happen.